In [2]:
import json

with open('documents-with-ids.json', 'rt') as f_in:
    documents = json.load(f_in)

In [1]:
from sentence_transformers import SentenceTransformer

In [3]:
model_name = 'multi-qa-MiniLM-L6-cos-v1'
model = SentenceTransformer(model_name)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/11.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/383 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [5]:
v = model.encode('I just discover the course. Can I still join?')

In [6]:
v.dot(v)

1.0

Remember to run the docker container first

In [10]:
from elasticsearch import Elasticsearch

es_client = Elasticsearch('http://localhost:9200') 

index_settings = {
    "settings": {
        "number_of_shards": 1,
        "number_of_replicas": 0
    },
    "mappings": {
        "properties": {
            "text": {"type": "text"},
            "section": {"type": "text"},
            "question": {"type": "text"},
            "course": {"type": "keyword"},
            "id": {"type": "keyword"},
            "question_vector": {
                "type": "dense_vector",
                "dims": 384,
                "index": True,
                "similarity": "cosine"
            },
            "text_vector": {
                "type": "dense_vector",
                "dims": 384,
                "index": True,
                "similarity": "cosine"
            },
            "question_text_vector": {
                "type": "dense_vector",
                "dims": 384,
                "index": True,
                "similarity": "cosine"
            },
        }
    }
}

index_name = "course-questions"

es_client.indices.delete(index=index_name, ignore_unavailable=True)
es_client.indices.create(index=index_name, body=index_settings)

ObjectApiResponse({'acknowledged': True, 'shards_acknowledged': True, 'index': 'course-questions'})

In [8]:
from tqdm.auto import tqdm

for doc in tqdm(documents):
    question = doc['question']
    text = doc['text']
    qt = question + ' ' + text

    doc['question_vector'] = model.encode(question)
    doc['text_vector'] = model.encode(text)
    doc['question_text_vector'] = model.encode(qt)

  0%|          | 0/948 [00:00<?, ?it/s]

In [11]:
for doc in tqdm(documents):
    es_client.index(index=index_name, document=doc)

  0%|          | 0/948 [00:00<?, ?it/s]

In [13]:
query = 'I just discovered the course. Can I still join it?'

In [14]:
v_q = model.encode(query)

In [20]:
def elastic_search_knn(field, vector, course):
    knn = {
        "field": field,
        "query_vector": vector,
        "k": 5,
        "num_candidates": 10000,
        "filter": {
            "term": {
                "course": course
            }
        }
    }

    search_query = {
        "knn": knn,
        "_source": ["text", "section", "question", "course", "id"]
    }

    es_results = es_client.search(
        index=index_name,
        body=search_query
    )
    
    result_docs = []
    
    for hit in es_results['hits']['hits']:
        result_docs.append(hit['_source'])

    return result_docs

In [22]:
def question_vector_knn(q):
    question = q['question']
    course = q['course']

    v_q = model.encode(question)
    return elastic_search_knn('question_vector', v_q, course)

In [31]:
def text_vector_knn(q):
    question = q['question']
    course = q['course']

    v_q = model.encode(question)
    return elastic_search_knn('text_vector', v_q, course)

In [35]:
def question_text_vector_knn(q):
    question = q['question']
    course = q['course']

    v_q = model.encode(question)
    return elastic_search_knn('question_text_vector', v_q, course)

In [23]:
import pandas as pd

In [24]:
df_ground_truth = pd.read_csv('ground-truth-data.csv')

In [25]:
df_ground_truth

question  \
0                           When does the course begin?   
1                    How can I get the course schedule?   
2             What is the link for course registration?   
3               How can I receive course announcements?   
4                    Where do I join the Slack channel?   
...                                                 ...   
4622  How can I resolve the Docker error 'invalid mo...   
4623  What should I do if I encounter an invalid mod...   
4624  What is the correct mounting path to use in Do...   
4625  Can you provide an example of a correct Docker...   
4626  How do I correct the mounting path error in Do...   

                         course  document  
0     data-engineering-zoomcamp  c02e79ef  
1     data-engineering-zoomcamp  c02e79ef  
2     data-engineering-zoomcamp  c02e79ef  
3     data-engineering-zoomcamp  c02e79ef  
4     data-engineering-zoomcamp  c02e79ef  
...                         ...       ...  
4622             mlops-zoomcamp  886d1617  
4623             mlops-zoomcamp  886d1617  
4624             mlops-zoomcamp  886d1617  
4625             mlops-zoomcamp  886d1617  
4626             mlops-zoomcamp  886d1617  

[4627 rows x 3 columns]

In [26]:
ground_truth = df_ground_truth.to_dict(orient='records')

## Evaluation metrics

- hit-rate (recall)
- Mean Reciprocal Rank (mrr)

In [27]:
def hit_rate(relevance_total):
    cnt = 0
    for line in relevance_total:
        if True in line:
            cnt += 1
    return cnt / len(relevance_total)

Mean reciprocal rank is similar to hit rate but it also looks at the position to calculate the metric, this improve the "accuracy" by using the postion

i.e:
if the position is 1 we add 1
if the position is 2 we add 1/2 = 0.5
if the position is 3 we add 1/3 = 0.333 and so on

The position would be the denominator and 1 will always be the numerator

In [28]:
def mmr(relevance_total):
    total_score = 0.0
    
    for line in relevance_total:
        for rank in range(len(line)):
            if line[rank] == True:
                total_score = total_score + 1/ (rank + 1)
    return total_score / len(relevance_total)
            

Let's do the same for min_search

In [29]:
import minsearch
# we add one more field than the first week ('id')
index = minsearch.Index(
    text_fields=['question', 'text', 'section'],
    keyword_fields=['course', 'id']
)
index.fit(documents)

In [30]:
def minsearch_search(query, course):
    boost = {'question': 3.0, 'section': 0.5}

    results = index.search(
        query=query,
        filter_dict={'course': course},
        boost_dict=boost,
        num_results=5
    )

    return results

Compare with ES results:

```
(0.7384914631510698, 0.6018730639002958)
```
    

In our results minsearch perform slightly better than elastic search

In [29]:
def evaluate(ground_truth, search_function):
    relevance_total = []

    for q in tqdm(ground_truth):
        doc_id = q['document']
        results = search_function(q)
        relevance = [d['id'] == doc_id for d in results]
        relevance_total.append(relevance)

    return {'hit_rate': hit_rate(relevance_total),
            'mmr': mmr(relevance_total)
           }

In [33]:
text_vector_knn(ground_truth[0])

[{'question': 'Course - When will the course start?',
  'course': 'data-engineering-zoomcamp',
  'section': 'General course-related questions',
  'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
  'id': 'c02e79ef'},
 {'question': 'Course - Can I follow the course after it finishes?',
  'course': 'data-engineering-zoomcamp',
  'section': 'General course-related questions',
  'text': 'Yes, we will keep all the materials after the course finishes, so you can follow the course at your own pace after it finishes.\nYou can also continue looking at the homeworks and continu

### Only question vector

### Only text vector

In [30]:
evaluate(ground_truth, question_vector_knn)

  0%|          | 0/4627 [00:00<?, ?it/s]

{'hit_rate': 0.7717743678409337, 'mmr': 0.6658273899574965}

In [34]:
evaluate(ground_truth, text_vector_knn)

  0%|          | 0/4627 [00:00<?, ?it/s]

{'hit_rate': 0.8275340393343419, 'mmr': 0.7052589871046767}

### combination between question and text vector

In [36]:
evaluate(ground_truth, question_text_vector_knn)

  0%|          | 0/4627 [00:00<?, ?it/s]

{'hit_rate': 0.9161443700021612, 'mmr': 0.823225992363663}

In this section we want to compare the three dense vector we use: 
- question_vector
- text_vector
- question_text_vector

In [39]:
def elastic_search_knn_combined(vector, course):
    search_query = {
        "size": 5,
        "query": {
            "bool": {
                "must": [
                    {
                        "script_score": {
                            "query": {
                                "term": {
                                    "course": course
                                }
                            },
                            "script": {
                                "source": """
                                    cosineSimilarity(params.query_vector, 'question_vector') + 
                                    cosineSimilarity(params.query_vector, 'text_vector') + 
                                    cosineSimilarity(params.query_vector, 'question_text_vector') + 
                                    1
                                """,
                                "params": {
                                    "query_vector": vector
                                }
                            }
                        }
                    }
                ],
                "filter": {
                    "term": {
                        "course": course
                    }
                }
            }
        },
        "_source": ["text", "section", "question", "course", "id"]
    }

    es_results = es_client.search(
        index=index_name,
        body=search_query
    )
    
    result_docs = []
    
    for hit in es_results['hits']['hits']:
        result_docs.append(hit['_source'])

    return result_docs

In [40]:
def vector_combined_knn(q):
    question = q['question']
    course = q['course']

    v_q = model.encode(question)

    return elastic_search_knn_combined(v_q, course)

evaluate(ground_truth, vector_combined_knn)

  0%|          | 0/4627 [00:00<?, ?it/s]

{'hit_rate': 0.9012318997190404, 'mmr': 0.8034003313882291}

Given the overall three results the best model is the one that combines question and text with {'hit_rate': 0.9161443700021612, 'mmr': 0.823225992363663}